File: src/rotating_the_box_1861/solution.py
from typing import List


class Solution:
    def rotateTheBox(self, boxGrid: List[List[str]]) -> List[List[str]]:
        m = len(boxGrid)
        n = len(boxGrid[0])

        for row in range(m):
            right_border = n - 1
            for col in range(n - 1, -1, -1):
                obj = boxGrid[row][col]
                if obj == "*":
                    right_border = col - 1
                elif boxGrid[row][col] == "#":
                    if col == right_border:
                        right_border -= 1
                        continue
                    boxGrid[row][col] = "."
                    boxGrid[row][right_border] = "#"
                    right_border -= 1

        result = [["." for _ in range(m)] for _ in range(n)]

        for row in range(m):
            for col in range(n):
                result[col][m - 1 - row] = boxGrid[row][col]

        return result


def test():
    solution = Solution()
    print(solution.rotateTheBox([["#", ".", "#"]]))
    print(solution.rotateTheBox([["#", ".", "*", "."], ["#", "#", "*", "."]]))
    print(
        solution.rotateTheBox(
            [
                ["#", "#", "*", ".", "*", "."],
                ["#", "#", "#", "*", ".", "."],
                ["#", "#", "#", ".", "#", "."],
            ]
        )
    )

File: src/rotating_the_box_1861/solution.rs
pub struct Solution;

impl Solution {
    pub fn rotate_the_box(box_grid: Vec<Vec<char>>) -> Vec<Vec<char>> {
        let (n, m) = (box_grid.len(), box_grid[0].len());
        let mut result_grid: Vec<Vec<char>> = vec![vec!['.'; n]; m];

        for (i, row) in box_grid.iter().enumerate() {
            let mut last = m - 1;
            for (j, ch) in row.iter().enumerate().rev() {
                match *ch {
                    '#' => {
                        result_grid[last][n - 1 - i] = '#';
                        last = last.saturating_sub(1);
                    }
                    '*' => {
                        result_grid[j][n - 1 - i] = '*';
                        last = j - 1;
                    }
                    _ => continue,
                }
            }
        }

        result_grid
    }
}

fn print_grid(grid: &Vec<Vec<char>>) {
    for row in grid {
        for &col in row {
            print!("{} ", col);
        }
        println!();
    }
    println!();
}

pub fn test() {
    let grid1 = vec![vec!['#', '.', '#']];
    print_grid(&Solution::rotate_the_box(grid1));

    let grid2 = vec![vec!['#', '.', '*', '.'], vec!['#', '#', '*', '.']];
    print_grid(&Solution::rotate_the_box(grid2));

    let grid3 = vec![
        vec!['#', '#', '*', '.', '*', '.'],
        vec!['#', '#', '#', '*', '.', '.'],
        vec!['#', '#', '#', '.', '#', '.'],
    ];
    print_grid(&Solution::rotate_the_box(grid3));
}
